File: client/src/pages/Document.js
// @flow

import React, { useCallback } from 'react';
import { BooleanIcon, LazyImage } from '@performant-software/semantic-components';
import { useTranslation } from 'react-i18next';
import { Header, Segment } from 'semantic-ui-react';
import _ from 'underscore';
import AttributesGrid from '../components/AttributesGrid';
import DocumentActions from '../components/DocumentActions';
import DocumentsService from '../services/Documents';
import RecordPage from '../components/RecordPage';
import useCurrentRecord from '../hooks/CurrentRecord';

const Document = () => {
  /**
   * Callback for loading the current document.
   *
   * @type {function(*): Promise<AxiosResponse<T>>|Promise<AxiosResponse<T>|unknown>|Promise<unknown>|*}
   */
  const onLoad = useCallback((id) => (
    DocumentsService
      .fetchOne(id)
      .then(({ data }) => data.document)
  ), []);

  const { item, loading } = useCurrentRecord(onLoad);
  const { t } = useTranslation();

  return (
    <RecordPage
      className='document'
      loading={loading}
      renderTitle={() => item.name}
    >
      { item && item.primary_attachment && (
        <RecordPage.Section
          className='image-container'
        >
          <LazyImage
            src={item.primary_attachment.file_url}
          />
        </RecordPage.Section>
      )}
      <RecordPage.Section>
        <AttributesGrid
          attributes={[{
            name: 'id',
            label: t('Common.labels.id')
          }, {
            name: 'name',
            label: t('Document.labels.name')
          }, {
            name: 'document_type',
            label: t('Document.labels.documentType'),
            qualification: {
              object: 'Document',
              group: 'Document Type'
            }
          }, {
            name: 'document_format',
            label: t('Document.labels.documentFormat'),
            qualification: {
              object: 'Document',
              group: 'Document Format'
            }
          }, {
            name: 'orientation',
            label: t('Document.labels.orientation'),
            qualification: {
              object: 'Document',
              group: 'Orientation (spine)'
            }
          }, {
            name: 'size',
            label: t('Document.labels.size'),
            qualification: {
              object: 'Document',
              group: 'Size'
            }
          }, {
            name: 'aperture',
            label: t('Document.labels.aperture'),
            qualification: {
              object: 'Document',
              group: 'Aperture'
            }
          }, {
            name: 'notes',
            label: t('Document.labels.notes')
          }]}
          item={item}
        />
      </RecordPage.Section>
      { !_.isEmpty(item?.actions) && (
        <RecordPage.Section
          title={t('Document.tabs.actions')}
        >
          <Segment>
            <DocumentActions
              items={item.actions}
            />
          </Segment>
        </RecordPage.Section>
      )}
      <RecordPage.Section
        title={t('Document.tabs.external')}
      >
        <Header
          content={t('Document.sections.binding')}
        />
        <AttributesGrid
          attributes={[{
            name: 'binding_type',
            label: t('Document.labels.bindingType'),
            qualification: {
              object: 'Document',
              group: 'Binding Type'
            }
          }, {
            name: 'binding_color',
            label: t('Document.labels.bindingColor'),
            qualification: {
              object: 'Document',
              group: 'Binding Color'
            }
          }, {
            name: 'sewing_supports_visible',
            label: t('Document.labels.sewingSupportsVisible'),
            renderValue: () => <BooleanIcon value={item.sewing_supports_visible} />
          }, {
            name: 'number_sewing_supports',
            label: t('Document.labels.numberSewingSupports')
          }, {
            name: 'spine_features',
            label: t('Document.labels.spineFeatures'),
            qualification: {
              object: 'Document',
              group: 'Spine Features'
            }
          }, {
            name: 'furniture',
            label: t('Document.labels.furniture'),
            qualification: {
              object: 'Document',
              group: 'Furniture'
            }
          }, {
            name: 'fastenings',
            label: t('Document.labels.fastenings'),
            qualification: {
              object: 'Document',
              group: 'Fastenings'
            }
          }, {
            name: 'number_fastenings',
            label: t('Document.labels.numberFastenings')
          }, {
            name: 'location_fastenings',
            label: t('Document.labels.locationsOfFastenings'),
            qualification: {
              object: 'Document',
              group: 'Location of Fastenings'
            }
          }, {
            name: 'inscriptions_on_binding',
            label: t('Document.labels.inscriptionsOnBinding'),
            renderValue: () => <BooleanIcon value={item.inscriptions_on_binding} />
          }, {
            name: 'inscription_text',
            label: t('Document.labels.inscriptionText')
          }, {
            name: 'binding_ornamentation',
            label: t('Document.labels.bindingOrnamentation'),
            qualification: {
              object: 'Document',
              group: 'Binding Ornamentation'
            }
          }, {
            name: 'binding_iconography',
            label: t('Document.labels.bindingIconography'),
            qualification: {
              object: 'Document',
              group: 'Iconography'
            }
          }]}
          item={item}
        />
        <Header
          content={t('Document.sections.endband')}
        />
        <AttributesGrid
          attributes={[{
            name: 'endband_present',
            label: t('Document.labels.endbandPresent'),
            renderValue: () => <BooleanIcon value={item.endband_present} />
          }, {
            name: 'endband_colors',
            label: t('Document.labels.endbandColors'),
            qualification: {
              object: 'Document',
              group: 'Color',
              formField: 'endband_colors'
            }
          }, {
            name: 'endband_style',
            label: t('Document.labels.endbandStyle'),
            qualification: {
              object: 'Document',
              group: 'Endband Style'
            }
          }]}
          item={item}
        />
        <Header
          content={t('Document.sections.foreEdges')}
        />
        <AttributesGrid
          attributes={[{
            name: 'binding_relationship_text_block',
            label: t('Document.labels.bindingRelationshipToTextBlock'),
            qualification: {
              object: 'Document',
              group: 'Binding Relationship to Text Block'
            }
          }, {
            name: 'decorated_fore_edges',
            label: t('Document.labels.decoratedForeEdges'),
            qualification: {
              object: 'Document',
              group: 'Decorated Fore-Edges'
            }
          }, {
            name: 'uncut_fore_edges',
            label: t('Document.labels.uncutForeEdges'),
            renderValue: () => <BooleanIcon value={item.uncut_fore_edges} />
          }, {
            name: 'fore_edges_color',
            label: t('Document.labels.colorOfForeEdges'),
            qualification: {
              object: 'Document',
              group: 'Color',
              formField: 'fore_edges_color'
            }
          }, {
            name: 'fore_edge_text',
            label: t('Document.labels.foreEdgeText')
          }]}
          item={item}
        />
        <Header
          content={t('Document.sections.bookmarks')}
        />
        <AttributesGrid
          attributes={[{
            name: 'bookmarks_registers',
            label: t('Document.labels.bookmarksRegisters')
          }, {
            name: 'bookmark_register_color',
            label: t('Document.labels.color'),
            qualification: {
              object: 'Document',
              group: 'Color',
              formField: 'bookmark_register_color'
            }
          }, {
            name: 'bookmark_register_style',
            label: t('Document.labels.bookmarkRegisterStyle'),
            qualification: {
              object: 'Document',
              group: 'Bookmark/Register Style'
            }
          }]}
          item={item}
        />
      </RecordPage.Section>
      <RecordPage.Section
        title={t('Document.tabs.internal')}
      >
        <Header
          content={t('Document.sections.layout')}
        />
        <AttributesGrid
          attributes={[{
            name: 'text_technology',
            label: t('Document.labels.textTechnology'),
            qualification: {
              object: 'Document',
              group: 'Text Technology'
            }
          }, {
            name: 'text_columns',
            label: t('Document.labels.textColumns')
          }, {
            name: 'page_contents',
            label: t('Document.labels.pageContents'),
            qualification: {
              object: 'Document',
              group: 'Page Contents'
            }
          }, {
            name: 'ruling',
            label: t('Document.labels.ruling'),
            renderValue: () => <BooleanIcon value={item.ruling} />
          }, {
            name: 'ruling_color',
            label: t('Document.labels.rulingColor'),
            qualification: {
              object: 'Document',
              group: 'Color',
              formField: 'ruling_color'
            }
          }, {
            name: 'rubrication',
            label: t('Document.labels.rubrication'),
            renderValue: () => <BooleanIcon value={item.rubrication} />
          }, {
            name: 'rubrication_color',
            label: t('Document.labels.rubricationColor'),
            qualification: {
              object: 'Document',
              group: 'Color',
              formField: 'rubrication_color'
            }
          }]}
          item={item}
        />
        <Header
          content={t('Document.sections.text')}
        />
        <AttributesGrid
          attributes={[{
            name: 'legibility',
            label: t('Document.labels.legibility'),
            qualification: {
              object: 'Document',
              group: 'Legibility'
            }
          }, {
            name: 'script',
            label: t('Document.labels.script'),
            qualification: {
              object: 'Document',
              group: 'Script'
            }
          }, {
            name: 'language',
            label: t('Document.labels.language'),
            qualification: {
              object: 'Document',
              group: 'Language'
            }
          }, {
            name: 'identity',
            label: t('Document.labels.identity')
          }, {
            name: 'transcription',
            label: t('Document.labels.transcriptionDiplomatic')
          }, {
            name: 'transcription_expanded',
            label: t('Document.labels.transcriptionExpanded')
          }, {
            name: 'transcription_translation',
            label: t('Document.labels.transcriptionTranslation')
          }, {
            name: 'simulated_script',
            label: t('Document.labels.simulatedScript'),
            qualification: {
              object: 'Document',
              group: 'Simulated Script'
            }
          }, {
            name: 'type_of_illumination',
            label: t('Document.labels.typeOfIllumination'),
            qualification: {
              object: 'Document',
              group: 'Type of Illumination'
            }
          }, {
            name: 'illumination_iconography',
            label: t('Document.labels.illuminationIconography'),
            qualification: {
              object: 'Document',
              group: 'Illumination Iconography',
              formField: 'illumination_iconography'
            }
          }]}
          item={item}
        />
      </RecordPage.Section>
    </RecordPage>
  );
};

export default Document;

File: app/models/search/document.rb
module Search
  module Document
    extend ActiveSupport::Concern

    class_methods do
      def published
        preload = {
          **::Document.primary_attachment_preload,
          actions: {
            qualifications: :value_list
          },
          visual_context: {
            **::VisualContext.primary_attachment_preload,
            physical_component: {
              **::PhysicalComponent.primary_attachment_preload,
              artwork: {
                **::Artwork.primary_attachment_preload,
                artwork_titles: {
                  qualifications: :value_list
                },
                participations: {
                  qualifications: :value_list,
                  person: {
                    qualifications: :value_list
                  }
                },
                locations: [:place, qualifications: :value_list],
                qualifications: :value_list
              },
            },
            qualifications: :value_list
          },
          qualifications: :value_list
        }

        preload(preload)
          .joins(visual_context: { physical_component: :artwork })
          .where(artworks: { published: true })
      end
    end

    included do
      # Includes
      include Base

      # Search index attributes
      search_attribute :name
      search_attribute :notes
      search_attribute :document_format, object: 'Document', group: 'Document Format', facet: true
      search_attribute :document_type, object: 'Document', group: 'Document Type', multiple: true, facet: true
      search_attribute :orientation, object: 'Document', group: 'Orientation (spine)', multiple: true, facet: true
      search_attribute :size, object: 'Document', group: 'Size', facet: true
      search_attribute :aperture, object: 'Document', group: 'Aperture', facet: true
      search_attribute :binding_type, object: 'Document', group: 'Binding Type', multiple: true, facet: true
      search_attribute :binding_color, object: 'Document', group: 'Color', form_field: 'binding_color', multiple: true, facet: true
      search_attribute :number_sewing_supports, facet: true
      search_attribute :spine_features, object: 'Document', group: 'Spine Features', multiple: true, facet: true
      search_attribute :furniture, object: 'Document', group: 'Furniture', multiple: true, facet: true
      search_attribute :fastenings, object: 'Document', group: 'Fastenings', multiple: true, facet: true
      search_attribute :number_fastenings, facet: true
      search_attribute :location_fastenings, object: 'Document', group: 'Location of Fastenings', multiple: true, facet: true
      search_attribute :inscriptions_on_binding, facet: true
      search_attribute :inscription_text
      search_attribute :binding_ornamentation, object: 'Document', group: 'Binding Ornamentation', multiple: true, facet: true
      search_attribute :binding_iconography, object: 'Document', group: 'Iconography', form_field: 'binding_iconography', multiple: true, facet: true
      search_attribute :endband_present, facet: true
      search_attribute :endband_colors, object: 'Document', group: 'Color', form_field: 'endband_colors', multiple: true, facet: true
      search_attribute :endband_style, object: 'Document', group: 'Endband Style', multiple: true, facet: true
      search_attribute :binding_relationship, object: 'Document', group: 'Binding Relationship to Text Block', multiple: true, facet: true
      search_attribute :decorated_fore_edges, object: 'Document', group: 'Decorated Fore-Edges', multiple: true, facet: true
      search_attribute :uncut_fore_edges, facet: true
      search_attribute :fore_edges_color, object: 'Document', group: 'Color', form_field: 'fore_edges_color', multiple: true, facet: true
      search_attribute :fore_edge_text
      search_attribute :bookmarks_registers, facet: true
      search_attribute :bookmark_register_color, object: 'Document', group: 'Color', form_field: 'bookmark_register_color', multiple: true, facet: true
      search_attribute :bookmark_style, object: 'Document', group: 'Bookmark/Register Style', multiple: true, facet: true
      search_attribute :text_technology, object: 'Document', group: 'Text Technology', multiple: true, facet: true
      search_attribute :text_columns, facet: true
      search_attribute :page_contents, object: 'Document', group: 'Page Contents', multiple: true, facet: true
      search_attribute :ruling, facet: true
      search_attribute :ruling_color, object: 'Document', group: 'Color', form_field: 'ruling_color', multiple: true, facet: true
      search_attribute :rubrication, facet: true
      search_attribute :rubrication_color, object: 'Document', group: 'Color', form_field: 'rubrication_color', multiple: true, facet: true
      search_attribute :legibility, object: 'Document', group: 'Legibility', multiple: true, facet: true
      search_attribute :script, object: 'Document', group: 'Script', multiple: true, facet: true
      search_attribute :language, object: 'Document', group: 'Language', multiple: true, facet: true
      search_attribute :identity
      search_attribute :transcription
      search_attribute :transcription_expanded
      search_attribute :transcription_translation
      search_attribute :simulated_script, object: 'Document', group: 'Simulated Script', multiple: true, facet: true
      search_attribute :illumination_type, object: 'Document', group: 'Type of Illumination', multiple: true, facet: true
      search_attribute :illumination_iconography, object: 'Document', group: 'Iconography', form_field: 'illumination_iconography', multiple: true, facet: true

      search_attribute(:image_url) do
        primary_attachment&.file_url
      end

      search_attribute(:thumbnail_url) do
        primary_attachment&.thumbnail_url
      end

      search_attribute(:artwork) do
        visual_context&.physical_component&.artwork&.to_search_json
      end

      search_attribute(:actions) do
        actions.map{ |action| action.to_search_json }
      end

      search_attribute(:physical_component) do
        visual_context&.physical_component&.to_search_json
      end

      search_attribute(:visual_context) do
        visual_context&.to_search_json
      end
    end
  end
end
